threescale/src/utils.rs

```diff
@@ -1,13 +1,58 @@
-use crate::structs::Application;
+use crate::structs::{ Application, PeriodWindow };
+use proxy_wasm::hostcalls::{ set_shared_data, get_shared_data};
+use log::info;
+use std::time::{ SystemTime, Duration };
 
 // Returns Application from shared data with CAS integer
 pub fn get_application_from_cache(key: &str) -> Option<(Application,u32)> {
-    None
+    match get_shared_data(&key).unwrap()
+    {
+        (bytes,cas) => {
+            match bytes {
+                Some(data) => Some((bincode::deserialize::<Application>(&data).unwrap(),cas.unwrap())),
+                None => return None,
+            }
+        },
+    } 
 }
 
-// Returns true on set failure
-pub fn set_application_to_cache(key: &str, app: &Application) -> bool {
-    let max_retries = 10; // We can also allow this, to be set from configuration
-    // Add random sleep in-between as well
-    true
+fn get_cas_from_cache(key: &str) -> Option<u32> {
+    match get_shared_data(&key).unwrap()
+    {
+        (_b, cas_o) => {
+            match cas_o {
+                Some(_c) => cas_o,
+                None => None,
+            }
+        }
+    }
+}
+
+// Returns false on set failure
+pub fn set_application_to_cache(key: &str, app: &Application, overwrite: bool, num_tries: Option<u32>) -> bool {
+    let mut cas = Some(0);
```

Comment:
Just passing in `None` is symmetric with `get_shared_data` (returns `None` if it is actually `0`) and behaves as if it was `Some(0)`.

Reply:
Made the suggest changes.